jax/interpreters/pxla.py

```diff
@@ -412,41 +412,94 @@ def _shard_abstract_array(size, axis: int, x):
   return x.update(shape=tuple_delete(x.shape, axis))
 shard_aval_handlers[ShapedArray] = _shard_abstract_array
 
-def aval_to_result_handler(sharding_spec: Optional[ShardingSpec],
-                           indices: Optional[Tuple[Index]],
-                           aval: core.AbstractValue) -> Callable[
-                               [List[xb.xla_client.Buffer]], Any]:
+MeshAxisName = Any
+"""
+ArrayMapping specifies how an ndarray should map to mesh axes.
+
+Note that the ordering is crucial for the cases when this mapping is non-injective
+(i.e. when multiple mesh axes map to the same positional axis). Then, the
+order of entries of the mapping determines a major-to-minor order on mesh axes,
+according to which chunks of the value along the repeated dimension will be assigned.
+
+For example, consider a mapping {'x': 1, 'y': 1} and a mesh with shape {'x': 2, 'y': 3}.
+The second dimension of the value would get chunked into 6 pieces, and assigned to the
+mesh in a way that treats 'y' as the fastest changing (minor) dimension. In this case,
+that would mean that a flat list of chunks would get assigned to a flattened list of
+mesh devices without any modifications. If the mapping was {'y': 1, 'x': 1}, then the
+mesh devices ndarray would have to be transposed before flattening and assignment.
+"""
+ArrayMapping = OrderedDictType[MeshAxisName, int]
+
+AxisResource = Tuple[Optional[Tuple[Any, ...]], ...]
+
+def array_mapping_to_axis_resources(array_mapping: ArrayMapping) -> AxisResource:
+  if not array_mapping:
+    return tuple()
+  max_index = array_mapping[max(array_mapping, key=array_mapping.get)]  # type: ignore
+  reverse_map = defaultdict(list)
+  for axis, index in array_mapping.items():
+    reverse_map[index].append(axis)
+  return tuple(
+      tuple(reverse_map[i]) if reverse_map[i] else None for i in range(max_index + 1)
+  )
+
+def aval_to_result_handler(
+    sharding_spec: Optional[ShardingSpec],
+    indices: Optional[Tuple[Index]],
+    aval: core.AbstractValue,
+    global_aval: Optional[ShapedArray] = None,
+    out_axis_resources: Optional[AxisResource] = None,
+    global_mesh = None,
```

Comment:
This seems very redundant. Can we simplify it a bit? From what I understand all you need is (1) sharding_spec, (2) global aval, (3) global mesh. The rest should be possible to infer from those.

Reply:
@skye said that inferring here should not be done because this is a performance critical function. So if things can be computed in MeshExecutable as much as possible is good.

And since all of that information is already available, I don't want to pass in sharding_spec and recreate out_axis_resources since its already available.

Reply:
I'm also hoping we can remove the older args if/when we always output GSDAs.